### DISCUSSION TITLE
Bringing the "framework" structure to neko

### DISCUSSION THREAD

**Discussion by timofeymukha** [ExtremeFLOW/neko#722] [Ideas]
Hello! I think it may be very good if we start working/discussing the abstract layer, which will underpin objects that "do something every timestep".  This is essentially how the KTH framework operates and there is a similar, very powerful, functionality in OpenFOAM. Basically, an object that has some sort of

* `init`, run at the beginning of the simulation
* `execute` that runs every timestep (maybe 2 of these, one at the beginning and one at the end of the timestep)
* `finalize`, which runs at the end of the simulation
*  Depending on how we eventually handle output control there can also be a `write`

What's a good name for such an object? OpenFOAM uses `functionObject`, but that is an abuse of a C++ concept.

We will probably need an extra object to hold a collection of the above, which will in the right place in `simulation` run a loop to trigger the above functions for each object it holds.

Thoughts?

**Comment by njansson** [comment]
Very good to get the discussion going!

So if I understand it correctly, we could achieve something similar by adding an abstract class (let's call it `object` in this reply), with deferred implementation of `init`, `execute` and `finalize`. 

Then we can have a list of `objects` either in the `simulation` or in `case` (I would prefer `case`, since we then have one object representing an entire simulation case), and at a given frequency, we iterate over all the objects and call `execute`, similar to how we do it for i.e. outputs in the sampler.

As for writing, I would rather than "add" the object to the samplers list of output, such that all I/O is going through one sampler (but each output can for example, have a different output frequency, as discussed in PR #720 )

**Comment by timofeymukha** [comment]
Yes, exactly something like that!

**Reply by njansson** [reply]
Great! This should be easy to add (most classes are already written in a similar fashion. However, the difficult part is what call it...maybe `NekoObject` (inspired by the infamous `vtkObject`)?

**Reply by timofeymukha** [reply]
@njansson `SimulationComponent`? Damn, this is hard :-). `NekoObject` sounds like some mega-ABC that everyone should inherit from (isn't `vtkObject` just that?)

**Reply by njansson** [reply]
Exactly `vtkObject` is the infamous mega-object...

`SimulationComponent` sounds more like it! Or what does other people think @MartinKarp @adampep @pschlatt1 ?

**Reply by MartinKarp** [reply]
Simulationcompnenent sounds good to me

**Reply by njansson** [reply]
> Simulationcompnenent sounds good to me

Let's go with `Simulationcomponent`! Anyone who wants to sketch it up?
